File: leetcode-top-interview-150/20.py
class Solution:
    def __init__(self):
        self.parentheses = {"(": ")", "{": "}", "[": "]"}

    def isValid(self, s: str) -> bool:
        return self.recursiveIsValid(s, [])

    def recursiveIsValid(self, s, stack) -> bool:
        if not s:
            return False

        cur = s[0]

        if cur in self.parentheses:
            if len(s) == 1:
                return False
            return self.recursiveIsValid(s[1:], stack + [cur])

        if not stack:
            return False

        if self.parentheses.get(stack[-1]) == cur:
            if len(s) == 1 and len(stack) == 1:
                return True
            else:
                return self.recursiveIsValid(s[1:], stack[:-1])

        return False

File: leetcode-top-100-liked/20.kt
class Solution {
    fun isValid(s: String): Boolean {
        var stack = Stack<Char>()

        var map = mapOf('(' to ')', '{' to '}', '[' to ']')

        for (char in s) {
            if(char in map.keys) {
                stack.push(char)
                continue
            }

            if(stack.size == 0) {
                return false
            }

            if(char != map[stack.pop()]) {
                return false
            }
        }
        
        if(stack.size == 0){
            return true
        }

        return false    

    }
}